### DISCUSSION TITLE
Nested Arguments

### DISCUSSION THREAD

**Discussion by jhurder-cvet** [IlyaMatsuev/Apex-GraphQL-Client#17] [Ideas]
Hi - me again :)

This question is around arguments. I am trying to build out an argument that has nested properties:

``` 
orders (customerLink: { ref: $customerRef })  { 
```
I tried building the argument out like this: 
```
if (m_params.CustomerId != null)
{            
     parentNode.withArgument('customerLink', '{ref: $customerRef}');
}
```
 However it ends up rendering the argument with quotes as the toString method determined that the argument was a String value: 

![image](https://user-images.githubusercontent.com/60743749/191350761-a8e6231c-e412-4b3c-aed7-94f471afe8e0.png)

So temporarily, I altered GraphQLArgument to support a new x_GraphQLArgument type in the toString() method:  
```
when x_GQLArgument {
    return  GraphQLConstants.BRACE_LEFT + ((GraphQLArgument)value).key + GraphQLConstants.COLON +
          GraphQLConstants.SPACE + ((GraphQLArgument)value) + GraphQLConstants.BRACE_RIGHT;
```
then, my code that calls looks like this:
```
if (m_params.CustomerId != null)
{            
     parentNode.withArgument('customerLink', 
        new GraphQLArgument('ref', '$customerRef'));
}
```

Alas - this has done the trick, the arguments are now created properly. Is this the best way to handle this situation? If so, is it okay to create a pull request for this work?

Thanks again for your continued support!

**Comment by IlyaMatsuev** [comment]
Hi!
As far as I understand you want to put an object as an argument, right? It renders the argument with quotes because you paste it as a string. If you want to put an object you should try something like this instead:

```java
if (m_params.CustomerId != null)
{
     parentNode.withArgument('customerLink', new Map<String, Object> { 'ref' => '$customerRef' });
}
```

I'm not 100% sure this is gonna work for your case but would be nice if you try it - I don't have an opportunity right now to check it.

If this doesn't work then I'd like to ask you for the full query that you're trying to do and create an issue so I can investigate it further.

**Reply by jhurder-cvet** [reply]
I tried this before as well, I ended up with "$customerRef" because the getArgumentTypeByValue() method evaluates `value instanceof String` to be true. Here's what I ended up with:

```
if (m_params.CustomerId != null)
{            
      parentNode.withArgument('customerLink', new Map<string, object> { 'ref' => '$customerRef'});
}
yields the following response:

query OrderQuery($count: Int = 10, $cursor: String = null, $ref: [String] = null, $status: [String] = null, $customerRef: String!)  {
  orders(first: $count, after: $cursor, ref: $ref, status: $status, customerLink: {ref:"$customerRef"})
```
I'll create an issue and post my full query for your review. 
Thanks -Jason

**Reply by IlyaMatsuev** [reply]
Okay, I see the problem. Thanks for the issue

**Comment by IlyaMatsuev** [comment]
Hey, I think this should not be the case anymore. Can you please check out [this release](https://github.com/IlyaMatsuev/Apex-GraphQL-Client/releases/tag/1.4)? Thanks

**Reply by jhurder-cvet** [reply]
Looking good - I tested it against my project and is working as expected. Thank you for the quick turnaround!
